File: run-fairy-tale-migration.py
#!/usr/bin/env python3
"""
Fairy Tale Migration Runner - Executes Migration 14 via API
Adds species_requirement, gender_requirement, age_requirement columns to fairy_tale_roles table
"""

import requests
import time
from pathlib import Path

# Configuration
API_BASE = "https://backend-2-production-3de1.up.railway.app"
MIGRATION_FILE = "backend/fairytales/migrations/14_add_role_matching_requirements.up.sql"

def read_migration_file():
    """Read the SQL migration file"""
    migration_path = Path(__file__).parent / MIGRATION_FILE

    if not migration_path.exists():
        raise FileNotFoundError(f"Migration file not found: {migration_path}")

    with open(migration_path, 'r', encoding='utf-8') as f:
        content = f.read()

    print(f"📄 Read migration file: {migration_path.name}")
    print(f"   Size: {len(content)} characters")

    return content

def execute_migration_via_api(sql_content):
    """Execute SQL migration via API endpoint"""

    # Split SQL into individual statements (by semicolon)
    statements = [s.strip() for s in sql_content.split(';') if s.strip() and not s.strip().startswith('--')]

    print(f"\n🔄 Executing {len(statements)} SQL statements...")

    # Execute each statement
    success_count = 0
    error_count = 0

    for idx, statement in enumerate(statements, 1):
        # Skip comments and empty statements
        if not statement or statement.startswith('--') or statement.startswith('/*'):
            continue

        # Log progress every 10 statements
        if idx % 10 == 0:
            print(f"   Progress: {idx}/{len(statements)} statements...")

        try:
            # Call the API endpoint (assuming you have one like this)
            # You may need to adjust the endpoint based on your actual API
            response = requests.post(
                f"{API_BASE}/fairytales/run-migration-sql",
                json={"sql": statement + ";"},  # Add semicolon back
                headers={"Content-Type": "application/json"},
                timeout=30
            )

            if response.status_code == 200:
                success_count += 1
            else:
                # Check if it's a "column already exists" error (acceptable)
                error_text = response.text.lower()
                if 'already exists' in error_text or 'duplicate' in error_text:
                    print(f"   ⚠️  Statement {idx}: Column already exists (skipping)")
                    success_count += 1
                else:
                    print(f"   ❌ Statement {idx} failed: {response.status_code} - {response.text[:100]}")
                    error_count += 1

            # Small delay to avoid overwhelming the API
            time.sleep(0.1)

        except Exception as e:
            print(f"   ❌ Statement {idx} error: {str(e)}")
            error_count += 1

    print(f"\n📊 Migration Results:")
    print(f"   ✅ Successful: {success_count}")
    print(f"   ❌ Failed: {error_count}")
    print(f"   📝 Total: {len(statements)}")

    return success_count, error_count

def verify_migration():
    """Verify that migration was successful by checking for new columns"""
    print(f"\n🔍 Verifying migration...")

    try:
        # Try to query fairy tale roles to see if new columns exist
        # This assumes you have an endpoint to inspect the table schema
        response = requests.get(
            f"{API_BASE}/fairytales/test-migration",
            timeout=10
        )

        if response.status_code == 200:
            data = response.json()
            if 'species_requirement' in str(data):
                print("   ✅ Migration verified: species_requirement column exists")
                return True
            else:
                print("   ⚠️  Warning: Could not verify species_requirement column")
                return False
        else:
            print(f"   ⚠️  Could not verify migration: {response.status_code}")
            return False

    except Exception as e:
        print(f"   ⚠️  Verification failed: {str(e)}")
        print("   💡 This is OK if the endpoint doesn't exist yet")
        return None

def main():
    print("=" * 80)
    print("🎭 FAIRY TALE MIGRATION 14: Character Matching Requirements")
    print("=" * 80)
    print()
    print("This migration adds the following columns to fairy_tale_roles:")
    print("  - species_requirement (human/animal/magical_creature/any)")
    print("  - gender_requirement (male/female/neutral/any)")
    print("  - age_requirement (child/adult/elder/any)")
    print("  - size_requirement (tiny/small/medium/large/giant/any)")
    print("  - social_class_requirement (royalty/nobility/craftsman/commoner/outcast/any)")
    print()
    print("This enables PRECISE character matching for fairy tale stories!")
    print()

    try:
        # Read migration file
        sql_content = read_migration_file()

        # Execute migration
        success, errors = execute_migration_via_api(sql_content)

        # Verify migration
        verified = verify_migration()

        # Final status
        print()
        print("=" * 80)
        if errors == 0:
            print("✅ SUCCESS! Migration completed without errors")
            print()
            print("Next steps:")
            print("  1. Import new characters from: Logs/new-characters-for-pool.json")
            print("  2. Test story generation with 'Klassische Märchen' genre")
            print("  3. Verify König is human (not Eichhörnchen!)")
        else:
            print(f"⚠️  PARTIAL SUCCESS: {success} succeeded, {errors} failed")
            print()
            print("Some statements failed - this may be OK if columns already exist")
            print("Check the error messages above for details")
        print("=" * 80)

    except Exception as e:
        print()
        print("=" * 80)
        print(f"❌ MIGRATION FAILED: {str(e)}")
        print("=" * 80)
        return 1

    return 0

if __name__ == "__main__":
    exit(main())

File: run-fairy-tale-migration.js
const fs = require('fs');

console.log('=' .repeat(80));
console.log('🎭 FAIRY TALE MIGRATION 14: Character Matching Requirements');
console.log('='.repeat(80));
console.log();

// Read migration file
const migrationPath = 'backend/fairytales/migrations/14_add_role_matching_requirements.up.sql';
const sqlContent = fs.readFileSync(migrationPath, 'utf8');

console.log(`📄 Read migration file: ${migrationPath}`);
console.log(`   Size: ${sqlContent.length} characters`);

// Split into statements
const statements = sqlContent
  .split(';')
  .map(s => s.trim())
  .filter(s => s && !s.startsWith('--') && s.length > 10);

console.log(`\n🔄 Found ${statements.length} SQL statements to execute`);
console.log('\nExecuting via API...\n');

const API_BASE = 'https://backend-2-production-3de1.up.railway.app';

async function executeStatements() {
  let successCount = 0;
  let errorCount = 0;

  for (let i = 0; i < statements.length; i++) {
    const statement = statements[i];

    if (i % 10 === 0 && i > 0) {
      console.log(`   Progress: ${i}/${statements.length} statements...`);
    }

    try {
      const response = await fetch(`${API_BASE}/fairytales/run-migration-sql`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          sql: statement + ';',
          migrationName: '14_add_role_matching_requirements'
        })
      });

      if (response.ok) {
        successCount++;
      } else {
        const errorText = await response.text();
        const lowerError = errorText.toLowerCase();

        // Check if it's an "already exists" error (acceptable)
        if (lowerError.includes('already exists') || lowerError.includes('duplicate')) {
          console.log(`   ⚠️  Statement ${i + 1}: Column already exists (skipping)`);
          successCount++;
        } else {
          console.log(`   ❌ Statement ${i + 1} failed: ${response.status}`);
          console.log(`      Error: ${errorText.substring(0, 100)}`);
          errorCount++;
        }
      }

      // Small delay
      await new Promise(resolve => setTimeout(resolve, 100));

    } catch (error) {
      console.log(`   ❌ Statement ${i + 1} error: ${error.message}`);
      errorCount++;
    }
  }

  console.log(`\n📊 Migration Results:`);
  console.log(`   ✅ Successful: ${successCount}`);
  console.log(`   ❌ Failed: ${errorCount}`);
  console.log(`   📝 Total: ${statements.length}`);

  console.log('\n' + '='.repeat(80));
  if (errorCount === 0) {
    console.log('✅ SUCCESS! Migration completed without errors');
    console.log();
    console.log('Next steps:');
    console.log('  1. Import new characters from: Logs/talea-characters-2025-11-19T12-41-27-184Z.json');
    console.log('  2. Test story generation with "Klassische Märchen" genre');
    console.log('  3. Verify König is human (not Eichhörnchen!)');
  } else {
    console.log(`⚠️  PARTIAL SUCCESS: ${successCount} succeeded, ${errorCount} failed`);
    console.log();
    console.log('Some statements failed - this may be OK if columns already exist');
  }
  console.log('='.repeat(80));
}

executeStatements().catch(err => {
  console.error('\n❌ MIGRATION FAILED:', err.message);
  process.exit(1);
});
